pageserver/src/metrics.rs

```diff
@@ -1233,117 +1233,167 @@ pub(crate) struct SmgrOpTimerInner {
+            execution_started_at: at,
+        }
+    }
 
-        // update metrics for execution latency
-        inner.global_execution_latency_histo.observe(execution);
+    /// For all but the first caller, this is a no-op.
+    /// The first callers receives Some, subsequent ones None.
+    pub(crate) fn observe_execution_end_flush_start(
+        &mut self,
+        at: Instant,
+    ) -> Option<SmgrOpFlushInProgress> {
+        // NB: unlike the other observe_* methods, this one take()s.
+        #[allow(clippy::question_mark)] // maintain similar code pattern.
+        let Some(mut inner) = self.0.take() else {
+            return None;
+        };
+        let SmgrOpTimerState::BatchedNowExecuting {
+            execution_started_at,
+        } = &mut inner.timings
+        else {
+            return None;
+        };
+        // update metrics
+        let execution = at - *execution_started_at;
+        inner
+            .global_execution_latency_histo
+            .observe(execution.as_secs_f64());
         if let Some(per_timeline_execution_latency_histo) =
             &inner.per_timeline_execution_latency_histo
         {
-            per_timeline_execution_latency_histo.observe(execution);
+            per_timeline_execution_latency_histo.observe(execution.as_secs_f64());
         }
 
-        Some((now, inner))
+        // state transition
+        inner.timings = SmgrOpTimerState::Flushing;
+
+        // return the flush in progress object which
+        // will do the remaining metrics updates
+        let SmgrOpTimerInner {
+            global_flush_in_progress_micros,
+            per_timeline_flush_in_progress_micros,
+            ..
+        } = inner;
+        Some(SmgrOpFlushInProgress {
+            flush_started_at: at,
+            global_micros: global_flush_in_progress_micros,
+            per_timeline_micros: per_timeline_flush_in_progress_micros,
+        })
     }
 }
 
+pub(crate) struct SmgrOpFlushInProgress {
+    flush_started_at: Instant,
+    global_micros: IntCounter,
+    per_timeline_micros: IntCounter,
+}
+
 impl Drop for SmgrOpTimer {
     fn drop(&mut self) {
-        self.smgr_op_end();
+        // In case of early drop, update any of the remaining metrics with
+        // observations so that (started,finished) counter pairs balance out
+        // and all counters on the latency path have the the same number of
+        // observations.
+        // It's technically lying and it would be better if each metric had
+        // a separate label or similar for cancelled requests.
+        // But we don't have that right now and counter pairs balancing
+        // out is useful when using the metrics in panels and whatnot.
+        let now = Instant::now();
+        self.observe_throttle_start(now);
+        self.observe_throttle_done(ThrottleResult::NotThrottled { end: now });
+        self.observe_execution_start(now);
+        self.observe_execution_end_flush_start(now);
```

Comment:
This could be misleading if we want to answer questions as "how much time on average did a request spent being throttled". Although, these early drop cases should be rare, so perhaps it doesn't matter.

Reply:
Yeah, as it says in the comment right above this quote.
It would be nicer to have a label in the _finished() counters for the outcome, e.g., `"ok", "err", "dropped"`.

The label value would be passed in as an argument to the `observe_*` functions.

But ENOTIME